### DISCUSSION TITLE
dat.gui, dat.guiVR.git

### DISCUSSION THREAD

**Discussion by sylwesterdigital** [felixmariotto/three-mesh-ui#56] [Q&A]
Are you planning to implement more UI components like we have with dat.gui?
Something like this?
https://github.com/dataarts/dat.guiVR.git

**Comment by felixmariotto** [comment]
Hi @sylwesterdigital, see [my reply on three.js discourse](https://discourse.threejs.org/t/dat-gui-for-webxr/21906/2).

I'm in for adding components like that ! But how do you want to control them ? three-mesh-ui purpose is strictly to offer XR user interfaces, not inputs. Or I could create a `Tickbox` and a `Slider` components, with methods such as `Tickbox.mark()`, but you would have to make the raycasting against the tickbox on your own, it would make your life easier for sure, but not as easy as dat.gui.

I can also create an example on how to use these elements and do the raycasting to use them, but like the [buttons example](https://three-mesh-ui.herokuapp.com/#interactive_button), the code for inputs would be part of the example, not of the library. Arguably this code could be copy-pasted, so it would not be much work for you if you want to skip this, but you are still free to use another kind of controls.

The reason why three-mesh-ui must keep controls out of its mission is because it can be used in a lot of different situations, commanding different inputs and controls.

**Reply by eviltik** [reply]
Hi

Agree with felix regarding controls.

So, a (new) project similar to dat.guiVR should be build "on top of" three-mesh-ui for controls. (xr, mouse, ..)  Another project, another repo.

I see 7.x.x roadmap, and it seem's more component will come. Slider and Tickbox is a good starting point. a basic ColorPicker is welcome too. 

Another widget should be a drop down, let's say a scrolling list. But here it's more complicated. How to handle scrolling if tree-mesh-ui don't handle controllers directly ?

**Reply by swingingtom** [reply]
The latest idea Im aware of : 

Starting with 7.x.x, three-mesh-ui will provide more input/interactive/behaviour components as **"examples"**. 
_The raycaster will be one of those._

Goals behind components : 

- Be use and reuse to build more complex and interactive (gui way) examples/samples/demo
- Be use as base for other components
- Be use and reuse by anyone on any project ( as-is : where they fit )
- Be documented samples to provide a guide to following to produce own component ( where they didn't fit as-is )

TL;DR those components will be provided as examples, in order to keep them out of the initial library scope.
